src/main/java/org/json/JSONObject.java

```diff
@@ -2886,4 +2889,33 @@ private static JSONException recursivelyDefinedObjectException(String key) {
             "JavaBean object contains recursively defined member variable of key " + quote(key)
         );
     }
+
+    /**
+     * For a prospective number, remove the leading zeros
+     * @param value prospective number
+     * @return number without leading zeros
+     */
+    private static String removeLeadingZerosOfNumber(String value){
+        char[] chars = value.toCharArray();
+        int leftMostUnsignedIndex = 0;
+        if (chars[0] == '-'){
+            leftMostUnsignedIndex = 1;
+        }
+        int firstNonZeroCharIndex = -1;
+        for (int i=leftMostUnsignedIndex;i<value.length();i++){
+            if (chars[i] != '0'){
+                firstNonZeroCharIndex = i;
+                break;
+            }
+        }
+        if (firstNonZeroCharIndex == -1){
+            return value;
+        }
+        StringBuilder result = new StringBuilder();
```

Comment:
The StringBuilder should only be used if `leftMostUnsignedIndex` is `1` (or `true`). Otherwise, just the result of the `substring` can be used.

Really, I'm not sure `StringBuilder` should be used directly at all here. The compiler will usually generate that for us by doing a simple concat. The following 2 samples should output almost identical byte code, however, the first is much easier to read:

```java
if (leftMostUnsignedIndex) {
  return '-' + value.substring(firstNonZeroCharIndex);
} else {
  return value.substring(firstNonZeroCharIndex);
}
```

```java
if (leftMostUnsignedIndex) {
  String val = value.substring(firstNonZeroCharIndex);
  StringBuilder tmp = new StringBuilder(val.length+1)
                             .append('-').append(val);
  return tmp.toString();
} else {
  return value.substring(firstNonZeroCharIndex);
}
```

Reply:
I also am not a fan of the `toCharArray`. I don't think we need to make a copy of the string to do the evaluation. I think a solution more like this is in line with performance and readability.

```java
/**
 * For a prospective number, remove the leading zeros
 * @param value prospective number.
 * @return number without leading zeros
 */
private static String removeLeadingZerosOfNumber(String value){
    if(value == null || value.length() == 0){
        return value;
    }

    boolean leftMostUnsignedIndex = value.charAt(0) == '-';
    int firstNonZeroCharIndex=-1;
    char firstNonZeroChar = 0;

    for(int i=leftMostUnsignedIndex ? 1 : 0; i < value.length(); i++){
    	firstNonZeroChar = value.charAt(i);
        if(firstNonZeroChar != '0') {
        	firstNonZeroCharIndex = i;
            break;
        }
    }

	if(firstNonZeroCharIndex < 0) {
    	return value;
    }
        
    if(firstNonZeroChar == '.' && value.charAt(firstNonZeroCharIndex-1)=='0') {
        firstNonZeroCharIndex--;
    }

    if (leftMostUnsignedIndex) {
        return '-' + value.substring(firstNonZeroCharIndex);
    } else {
        return value.substring(firstNonZeroCharIndex);
    }
}
```

Reply:
Agreed. toCharArray was unnecessary. This method has been cleaned.
Also, there was another fatal error in initial string parsing to number logic, that prevented any decimal number starting with '.' to be considered as decimal number, that has been fixed as well.